File: web/src/lib/sharedContact.ts
import type { SocialField } from "../components/SocialIcon";
import { SOCIAL_FIELDS, resolveSocialUrl } from "../components/SocialIcon";
import type { UserProfile } from "../api/types";

export const SOCIAL_LABELS: Record<SocialField, string> = {
  website: "Website",
  instagram: "Instagram",
  tiktok: "TikTok",
  youtube: "YouTube",
  facebook: "Facebook",
  twitter: "X / Twitter",
  app_store: "App Store",
  google_play: "Google Play",
  github: "GitHub",
  discord: "Discord",
};

export const CHANNEL_LABELS: Record<string, string> = {
  email: "Email",
  phone: "Phone",
  ...SOCIAL_LABELS,
};

export interface Channel {
  key: string;
  label: string;
  value: string;
}

export function availableChannels(profile: UserProfile): Channel[] {
  const channels: Channel[] = [{ key: "email", label: "Email", value: profile.email }];
  if (profile.phone) channels.push({ key: "phone", label: "Phone", value: profile.phone });
  for (const field of SOCIAL_FIELDS) {
    const value = profile.social_links?.[field];
    if (value) channels.push({ key: field, label: SOCIAL_LABELS[field], value });
  }
  return channels;
}

export function channelHref(key: string, value: string): string | null {
  if (key === "email") return `mailto:${value}`;
  if (key === "phone") return `tel:${value}`;
  if ((SOCIAL_FIELDS as string[]).includes(key)) {
    return resolveSocialUrl(key as SocialField, value);
  }
  return null;
}

File: api/app/services/shared_contact.py
from app.models.user import User

SOCIAL_FIELDS = (
    "website",
    "instagram",
    "tiktok",
    "youtube",
    "facebook",
    "twitter",
    "app_store",
    "google_play",
    "github",
    "discord",
)


def resolve_shared_contact(user: User, shared_fields: list[str]) -> dict[str, str]:
    """Filters a user's contact info down to the channels they've opted to share,
    e.g. with a project team or an idea's other interested builders."""
    available: dict[str, str | None] = {"email": user.email, "phone": user.phone}
    if user.social_links:
        for field in SOCIAL_FIELDS:
            available[field] = user.social_links.get(field)

    shared_contact: dict[str, str] = {}
    for field in shared_fields:
        value = available.get(field)
        if value:
            shared_contact[field] = value
    return shared_contact
